### PR TITLE
TDL-15031: Add addressable list fields (Copy of PR #34)

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-15031: #34 opened - Add addressable list fields (Copy of PR #34)
- Added addressable list fields in the Contacts stream. Copy of PR #34 

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_intercom/transform.py [hunk 1]
```diff
@@ -54,6 +55,11 @@ def transform_json(this_json, stream_name, data_key):
     elif stream_name == 'conversation_parts':
         denested_json = transform_conversation_parts(new_json, data_key)
         new_json = denested_json
+    elif stream_name == 'contacts':
+        # De-nest the addressable list fields in 'Contacts' stream
+        adressable_list_nodes = ['companies', 'notes', 'tags']
```
Comment (umangagrawal-crest) [inline tap_intercom/transform.py:60]: Remove 'notes' from the list

File: tap_intercom/streams.py [hunk 2]
```diff
@@ -533,9 +550,52 @@ class Contacts(IncrementalStream):
...
+                data = record.get(addressable_list_field)
+                paging = True
+                next_page = None
+                endpoint = data.get('url')
+
+                # Do not do the API call if we have 0 records
+                if not data.get('total_count') > 0:
+                    continue
+
+                # Do not do the API call if we have greater than 10 records ie. 'has_more' field is 'False'
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:580]:
```suggestion
                # Do not do the API call if we have less than 10 records ie. the 'has_more' field is 'False'
```

File: tap_intercom/transform.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_intercom/transform.py]: Removed notes schema.

File: tap_intercom/streams.py [hunk 3]
```diff
@@ -533,9 +550,52 @@ class Contacts(IncrementalStream):
...
+                if not data.get('has_more'):
+                    continue
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:582]: Combine these 2 conditions.

File: tap_intercom/streams.py [hunk 4]
```diff
@@ -533,9 +550,52 @@ class Contacts(IncrementalStream):
...
+        params = {
+            'display_as': 'plaintext',
+            'per_page': 60 # addressable_list endpoints have a different max page size in Intercom's API v2.0
+        }
 
+        for record in contact_list.get(self.data_key):
+            for addressable_list_field in self.addressable_list_fields:
+                # Do not do the API call to get addressable fields if the field is not selected
+                if not metadata.get(('properties', addressable_list_field), {}).get('selected'):
+                    continue
```
Comment (savan-chovatiya) [inline tap_intercom/streams.py:567]: Also, combine this condition with the other 2 conditions. Keep all `continue` related conditions together.

File: tap_intercom/streams.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Merged the `continue` conditions.

File: tap_intercom/streams.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Updated comment.

File: tap_intercom/streams.py (hunk 4 above)
Comment (harshpatel4crest) [reply tap_intercom/streams.py]: Merged the `continue` conditions.

File: tests/test_intercom_bookmarks.py [hunk 5]
```diff
@@ -109,17 +111,6 @@ def test_run(self):
...
-        conn_id = connections.ensure_connection(self, original_properties=False)
-
-        # run check mode
-        found_catalogs = self.run_and_verify_check_mode(conn_id)
-
-        # table and field selection
-        test_catalogs_2_all_fields = [catalog for catalog in found_catalogs
-                                      if catalog.get('tap_stream_id') in expected_streams]
-        self.perform_and_verify_table_and_field_selection(conn_id, test_catalogs_2_all_fields, select_all_fields=True)
-
```
Comment (savan-chovatiya) [inline tests/test_intercom_bookmarks.py:122]: Any reason to remove this field selection and check mode for the second sync?

File: tests/test_intercom_bookmarks.py [hunk 6]
```diff
@@ -67,21 +67,23 @@ def calculated_states_by_stream(self, current_state):
     def test_run(self):
         # Created card for untestable/unstable streams.
         # FIX CARD: https://jira.talendforge.org/browse/TDL-17035
-        untestable_streams = {"companies", "segments", "company_segments", "conversation_parts", "conversations"}
-        expected_streams =  self.expected_streams().difference(untestable_streams)
+        untestable_streams = {"companies", "company_segments", "conversation_parts", "conversations"}
+        # Contacts stream does 3 API calls for addressable list fields, [notes, companies, tags]
+        # This cause the build to run more than 3 hrs, thus skipping this stream
+        stream_to_skip = {"contacts"}
```
Comment (dbshah1212) [inline tests/test_intercom_bookmarks.py:73]:
```suggestion
        streams_to_skip = {"contacts"}
```

File: tests/test_intercom_bookmarks.py (hunk 5 above)
Comment (harshpatel4crest) [reply tests/test_intercom_bookmarks.py:122]: The 2nd sync was not running with the state file. After `menagerie.set_state(conn_id, new_states)` the conn_id is getting overridden and the 2nd sync will run with the start date. Thus, updated the code to create only 1 connection id and run 2nd sync with the state file.

**Review by savan-chovatiya — APPROVED** [review]

File: tests/test_intercom_bookmarks.py (hunk 6 above)
Comment (harshpatel4crest) [reply tests/test_intercom_bookmarks.py]: Updated the variable name.

**Review by dbshah1212 — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Pull request merged**